server/src/main/java/io/crate/replication/logical/MetadataTracker.java

```diff
@@ -435,6 +449,73 @@ private static void subscribeToNewRelations(String subscriptionName,
         );
     }
 
+    private ClusterState removeDroppedTablesOrPartitions(String subscriptionName,
+                                                         Subscription subscription,
+                                                         ClusterState subscriberClusterState,
+                                                         ClusterState publisherClusterState) {
+        HashSet<RelationName> relationsToUpdate = new HashSet<>();
+        ArrayList<String> indicesToRemove = new ArrayList<>();
+        ArrayList<String> templatesToRemove = new ArrayList<>();
+        for (var relationName : subscription.relations().keySet()) {
+            var possibleTemplateName = PartitionName.templateName(relationName.schema(), relationName.name());
+            var isPartitioned = subscriberClusterState.metadata().templates().get(possibleTemplateName) != null;
+            if (isPartitioned
+                && publisherClusterState.metadata().templates().get(possibleTemplateName) == null) {
+                templatesToRemove.add(possibleTemplateName);
+                relationsToUpdate.add(relationName);
+            }
+
+            var concreteIndices = indexNameExpressionResolver.concreteIndexNames(
+                subscriberClusterState,
+                IndicesOptions.lenientExpand(),
+                relationName.indexNameOrAlias()
+            );
+            for (var concreteIndex : concreteIndices) {
+                if (publisherClusterState.metadata().hasIndex(concreteIndex) == false) {
+                    indicesToRemove.add(concreteIndex);
+                    if (isPartitioned == false) {
+                        relationsToUpdate.add(relationName);
+                    }
+                }
+            }
+        }
+
+        var updatedClusterState = subscriberClusterState;
+        if (templatesToRemove.isEmpty() == false) {
+            var newMetadataBuilder = Metadata.builder(subscriberClusterState.metadata());
+            for (var templateName : templatesToRemove) {
+                newMetadataBuilder.removeTemplate(templateName);
+            }
+            updatedClusterState = ClusterState.builder(subscriberClusterState).metadata(newMetadataBuilder).build();
+        }
+
+        var future = CompletableFuture.completedFuture(true);
+        if (indicesToRemove.isEmpty() == false) {
+            var request = new DeleteIndexRequest(indicesToRemove.toArray(new String[0]));
+            request.indicesOptions(IndicesOptions.lenientExpand());
```

Comment:
`indicesToRemove` are already concrete indices, is `lenientExpand()` required here?